File: 백준/Silver/16926. 배열 돌리기 1/배열 돌리기 1.py
import sys
input = sys.stdin.readline

def rotate(sr, er, sc, ec, rot):
    ccnt, rcnt = ec - sc + 1, er - sr + 1
    if ccnt == 1:
        cnt = rcnt
    elif rcnt == 1:
        cnt = ccnt
    else:
        cnt = ccnt * 2 + rcnt * 2 - 4
    rot %= cnt

    tmp = arr[sr][sc:ec + 1]
    for i in range(sr + 1, er + 1):
        tmp.append(arr[i][ec])
    tmp += arr[er][sc:ec][::-1]
    for i in range(er - 1, sr, -1):
        tmp.append(arr[i][sc])
    tmp = tmp[rot:] + tmp[:rot]

    idx = 0
    for i in range(sc, ec + 1):
        ans[sr][i] = tmp[idx]
        idx += 1
    for i in range(sr + 1, er + 1):
        ans[i][ec] = tmp[idx]
        idx += 1
    for i in range(ec - 1, sc - 1, -1):
        ans[er][i] = tmp[idx]
        idx += 1
    for i in range(er - 1, sr, -1):
        ans[i][sc] = tmp[idx]
        idx += 1

N, M, R = map(int, input().split())
arr = [[*map(int, input().split())] for _ in range(N)]
ans = [[0] * M for _ in range(N)]
for i in range(min(N, M) // 2):
    rotate(i, N - 1 - i, i, M - 1 - i, R)
for a in ans:
    print(*a)

File: 백준/Silver/16926. 배열 돌리기 1/배열 돌리기 1.java
import java.io.*;
import java.util.*;

public class Main {
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		StringBuilder sb = new StringBuilder();
		int N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int R = Integer.parseInt(st.nextToken());
		int[][] src = new int[N][M], desc = new int[N][M];
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < M; j++)
				src[i][j] = Integer.parseInt(st.nextToken());
		}
		
		int cnt = Math.min(N, M) / 2;
		for (int i = 0; i < cnt; i++) {
			rotate(desc, src, i, N - i, i, M - i, R);
		}
		
		for (int i = 0; i < N; i++) {
			for (int j = 0; j < M; j++) {
				sb.append(desc[i][j]).append(" ");
			}
			sb.append("\n");
		}
		System.out.println(sb);
	}
	
	static void rotate(int[][] desc, int[][] src, int sr, int er, int sc, int ec, int R) {
		int rcnt = er - sr;
		int ccnt = ec - sc;
		
		if (rcnt == 1) R %= ccnt;
		else if (ccnt == 1) R %= rcnt;
		else R %= (rcnt * 2 + ccnt * 2 - 4);
		
		List<Integer> arr = new ArrayList<>();
		for (int i = sc; i < ec; i++)
			arr.add(src[sr][i]);
		for (int i = sr + 1; i < er; i++)
			arr.add(src[i][ec - 1]);
		for (int i = ec - 2; i >= sc; i--)
			arr.add(src[er - 1][i]);
		for (int i = er - 2; i > sr; i--)
			arr.add(src[i][sc]);
		
		Collections.rotate(arr, -R);
		
		int mCnt = 0;
		for (int i = sc; i < ec; i++)
			desc[sr][i] = arr.get(mCnt++);
		for (int i = sr + 1; i < er; i++)
			desc[i][ec - 1] = arr.get(mCnt++);
		for (int i = ec - 2; i >= sc; i--)
			desc[er - 1][i] = arr.get(mCnt++);
		for (int i = er - 2; i > sr; i--)
			desc[i][sc] = arr.get(mCnt++);
	}
	
}